pallets/dispenser/src/lib.rs

```diff
@@ -0,0 +1,390 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+use alloc::{string::String, vec};
+
+use alloy_primitives::U256;
+use alloy_sol_types::{sol, SolCall};
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::pallet_prelude::*;
+use frame_support::traits::fungibles::Inspect;
+use frame_support::traits::{fungibles::Mutate, tokens::Preservation, Currency};
+use frame_support::PalletId;
+use frame_support::{dispatch::DispatchResult, BoundedVec};
+use frame_system::pallet_prelude::*;
+use sp_core::H160;
+use sp_std::vec::Vec;
+
+#[cfg(feature = "runtime-benchmarks")]
+pub mod benchmarking;
+pub mod weights;
+
+#[cfg(test)]
+mod tests;
+
+pub use pallet::*;
+
+sol! {
+	#[sol(abi)]
+	interface IGasFaucet {
+		function fund(address to, uint256 amount) external;
+	}
+}
+
+pub type Balance = u128;
+pub type AssetId = u32;
+
+pub type EvmAddress = [u8; 20];
+
+pub type BalanceOf<T> =
+	<<T as pallet_signet::Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+#[derive(Encode, Decode, TypeInfo, Clone, Debug, PartialEq)]
+pub struct EvmTransactionParams {
+	pub value: u128,
+	pub gas_limit: u64,
+	pub max_fee_per_gas: u128,
+	pub max_priority_fee_per_gas: u128,
+	pub nonce: u64,
+	pub chain_id: u64,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use sp_runtime::traits::AccountIdConversion;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_signet::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		type Currency: Mutate<Self::AccountId, AssetId = AssetId, Balance = Balance>;
+
+		#[pallet::constant]
+		type MinimumRequestAmount: Get<u128>;
+		#[pallet::constant]
+		type MaxDispenseAmount: Get<u128>;
+		#[pallet::constant]
+		type DispenserFee: Get<u128>;
+
+		#[pallet::constant]
+		type FeeAsset: Get<AssetId>;
+		#[pallet::constant]
+		type FaucetAsset: Get<AssetId>;
+
+		#[pallet::constant]
+		type TreasuryAddress: Get<Self::AccountId>;
```

Comment:
let's rename this to FeeDestination. it might not be treasury, so let's use some generic term